Improving documentation.

useredsa opened the pull request:
Solves #24.

useredsa commented:
Any comments?

alexherbo2 commented:
@useredsa I do not have much time to review.

alexherbo2 commented on README.md:
```diff
@@ -7,6 +7,21 @@ Leverage the client-server architecture of [Kakoune] to connect programs to clie
 [![YouTube Play Button](https://www.iconfinder.com/icons/317714/download/png/16)](https://youtube.com/playlist?list=PLdr-HcjEDx_k-Y_9uSV0YAUCNHzqHjmz3) · [connect.kak](https://youtube.com/playlist?list=PLdr-HcjEDx_k-Y_9uSV0YAUCNHzqHjmz3)
 
+The objective of connect.kak is
+to sinchronize external applications with kakoune clients easily.
```
There is still a typo for “synchronize”.

useredsa replied:
I had solved that, but I think I reverted it back to normal with a bunch of ammends and rebase :disappointed:.

alexherbo2 commented on README.md:
```diff
@@ -7,6 +7,21 @@ Leverage the client-server architecture of [Kakoune] to connect programs to clie
+
+[modules]: rc/modules
+[commands]: rc/paths/commands
+[:batch]: https://discuss.kakoune.com/t/batch-processing-with-kakoune/620/5
```
Can you reference as “batch.kak” and the [project page][batch.kak]?

[batch.kak]: https://github.com/alexherbo2/batch.kak

alexherbo2 commented on docs/Recipes.md:
```diff
@@ -0,0 +1,44 @@
+
+[:batch]: https://github.com/alexherbo2/batch
+
+## Example configuration using [`plug.kak`][plug.kak]
```
I prefer to not link to plug.kak.

alexherbo2 commented on docs/Integration.md:
```diff
@@ -0,0 +1,46 @@
+(useful for applications that check the return value of the editor, like git).
+
+[commands]:   ../rc/connect/commands/
+[:attach]:    ../rc/connect/commands/:attach
```
You can reference as `` `:attach` `` and simplify the link references.

useredsa replied:
Do you mean replacing ``[`:[a]ttach`][:attach]`` with

``[`:attach`][:attach]``,

``[:attach]``,

or something else?

useredsa replied:
Or ``[`:[a]ttach`]``and later the ref with backticks and brackets too.

alexherbo2 replied:
``[`:attach`]``, as it is a code command.

alexherbo2 commented on docs/Integration.md:
```diff
@@ -0,0 +1,46 @@
+[:ls]:        ../rc/connect/commands/:ls
+[:send]:      ../rc/connect/commands/:send
+
+[yank-ring.kak]: https://github.com/alexherbo2/yank-ring.kak/blob/master/rc/yank-ring.kak
```
Prefer link to the home page.

alexherbo2 commented on docs/Recipes.md:
```diff
@@ -0,0 +1,44 @@
+# Recipes
+
+## Appending Kakoune's shell prompt to the usual prompt
```
Prefer apostrophes (’) everywhere.

alexherbo2 commented on docs/FAQ.md:
```diff
@@ -0,0 +1,72 @@
+to start a shell connected to that client's server.
+The file is automatically erased when it's sourced.
+
+Note 1: The kakoune server is deleted if you detach the last client and
```
Can you emphasis as `**Note 1**:` and capitalize Kakoune?

alexherbo2 commented on README.md:
```diff
@@ -180,7 +218,13 @@ PS1='$(~/.local/share/kak/connect/prompt) $ '
+- [Recipes]
+- [Integration with other tools][Integration]
+ 
+[Commands]:    rc/connect/commands
```
Prefer without alignment.

alexherbo2 commented on README.md:
````diff
@@ -155,18 +195,16 @@ set-option global connect_environment %{
 PS1='$(~/.local/share/kak/connect/prompt) $ '
 ```
 
+[prompt]: https://wiki.archlinux.org/index.php/Bash/Prompt_customization
````
Prefer references with full title, like “Prompt customization”.

useredsa replied:
```
You can modify your shell [prompt][Prompt customization]
to notify you whenever you are connected to a session.

...

[Prompt customization]: https://wiki.archlinux.org/index.php/Bash/Prompt_customization
```

Resolve if alright.

alexherbo2 commented on README.md:
```diff
@@ -76,7 +89,9 @@ $ dolphin
-**Kakoune** – Same with [modules]:
+**Kakoune** – Use [dolphin's module] to do the same as example 3:
+
+[dolphin's module]: rc/modules/dolphin.kak
```
Not sure if it’s “Dolphin module”.

useredsa replied:
I'd say `Dolphin module` means the name of the module is "Dolphin", in which case we could write "the module dolphin", while `Dolphin's module` means the module is about Dolphin.

alexherbo2 commented on README.md:
```diff
@@ -7,6 +7,19 @@ Leverage the client-server architecture of [Kakoune] to connect programs to clie
+The objective of connect.kak is
+to sinchronize external applications with kakoune clients easily.
+A typical use case is opening a file browser and
+having it open the files in the kakoune client.
```
Capitalize Kakoune.

alexherbo2 commented on docs/Integration.md:
```diff
@@ -0,0 +1,46 @@
+- [`:get`][:get]: Gets the result of a Kakoune's `echo` command from the client.
+- [`:[s]end`][:send]: Evaluates the arguments in the client.
+
+----
```
Prefer three `---`.

alexherbo2 commented on docs/Integration.md:
```diff
@@ -0,0 +1,46 @@
+# Integration with other apps
+
+`connect.kak` is also framework for developing your own plugins.
```
Prefer without code when referencing the name of the project.

alexherbo2 commented on docs/Integration.md:
```diff
@@ -0,0 +1,46 @@
+
+The basic [commands] for plugin crafting are:
+
+- [`:get`][:get]: Gets the result of a Kakoune's `echo` command from the client.
```
You can link to `:get` and `:send` (with the backticks).

alexherbo2 commented on docs/FAQ.md:
```diff
@@ -0,0 +1,72 @@
+Note 1: The kakoune server is deleted if you detach the last client and
+you didn't start it in daemon mode.
+
+Note 2: The implementation uses kakoune's `quit!` command instead of `quit`.
```
I want to quit, but I do not want to kill the server if it is not in daemon mode.

If you know a clean solution without complexifying the code.

useredsa replied:
There's not a way to transform the session into a headless session. If there was a way to query if it's a headless session or not, we could at least `quit` or `quit!`.

alexherbo2 commented on docs/FAQ.md:
````diff
@@ -0,0 +1,72 @@
+@kanto $ :a█
+```
+
+If you enter a number i, it will try to connect to the i-th entry.
````
Shouldn’t it be `_n_` and `_n_ᵗʰ`?

useredsa replied:
_n_ usually stands for natural and _i_ for integer. In this case, the code checks it's an integer. But I can change it to _n_ if you prefer.

Adding italics in any case.

useredsa replied:
i or  n then?

alexherbo2 replied:
What is the more fitting.

alexherbo2 commented on docs/FAQ.md:
```diff
@@ -0,0 +1,72 @@
+
+Apart from running a connected shell, you can use `kak-shell` to run
+commands in the context of the session.
+For example, to print the current working directory of the server, you can run
```
Is `pwd` useful?  `kak-shell :attach` looks more useful.

useredsa replied:
I prefer to put that inside a recipe, because it involves more things (it spawns a session as daemon mode if it doesn't exist, it's not the same as kak -c) and it's a nice tip.

`kak-shell tig`?

alexherbo2 commented on docs/Recipes.md:
```diff
@@ -0,0 +1,44 @@
+
+## [`:batch`][:batch] Integration to rename and drop buffers from the buffer list
+
+`~/.config/kak/connect/commands/:batch`
```
There is a new option for plugins to register their paths.  It can also be used by users to configure custom paths.

useredsa replied:
I'm going to remove it now that it is listed under integrations and the link to the code is in the readme.

alexherbo2 commented on README.md:
```diff
@@ -180,7 +218,13 @@ PS1='$(~/.local/share/kak/connect/prompt) $ '
+[Commands]:    rc/connect/commands
+[Aliases]:     rc/connect/aliases
+[Modules]:     rc/modules
+[FAQ]:         docs/FAQ.md
```
I prefer files to not be upcased.

alexherbo2 commented on docs/FAQ.md:
```diff
@@ -0,0 +1,72 @@
+while others spawn their own graphical window.
+Use `>` for the first type and `$` for the second.
+
+## What's the purpose of `&` or `:connect-detach`?
```
Not sure about `:connect-detach` and `connect-detach` in other places.

alexherbo2 commented on docs/Integration.md:
```diff
@@ -0,0 +1,46 @@
+
+## Interacting with Kakoune
+
+Integration with other applications usually comes from writting a small utility.
```
There is a typo for “writing”.

alexherbo2 commented on docs/Integration.md:
```diff
@@ -0,0 +1,46 @@
+The basic [commands] for plugin crafting are:
+
+- [`:get`][:get]: Gets the result of a Kakoune's `echo` command from the client.
+- [`:[s]end`][:send]: Evaluates the arguments in the client.
```
Sends commands to the client.

alexherbo2 commented:
@useredsa Looks good.  I haven’t finished the review.

alexherbo2 commented on README.md:
```diff
@@ -7,6 +7,19 @@ Leverage the client-server architecture of [Kakoune] to connect programs to clie
+
+connect.kak provides basic [commands] to interact with the connected client interactively
+or to write your own scripts
+(check [`:batch`][:batch], which is an integration crafted from the rest of the commands)
```
Can you simplify the reference as ``[`:batch`]``?

useredsa replied:
I had changed it (though I haven't pushed yet) to "batch.kak" and a link to alexherbo2/bath.kak, instead of the source. I can link to the source in the plugin, and display it as `` `:batch` `` then

alexherbo2 replied:
ouki

When done, could you squash the commits into a single one?

useredsa replied:
> ouki
> 
> When done, could you squash the commits into a single one?

Sure, I remember you prefer it squashed.

useredsa replied:
Done

useredsa committed:
Improved documentation.

Extended the plugin description and the examples.
Created a docs folder with the files faq.md, integration.md and recipes.md.

alexherbo2 commented:
@useredsa Tell me when it’s ready.

I added/renamed some commands in the meantime.  Maybe you want to update.

useredsa commented:
> @useredsa Tell me when it’s ready.
> 
> I added/renamed some commands in the meantime. Maybe you want to update.

It is ready right now.

Note that I moved ide and cd behaviour to recipes.

alexherbo2 commented on docs/faq.md:
```diff
@@ -0,0 +1,63 @@
+**Note 1**: The Kakoune server is deleted if you detach the last client and
+you didn´t start it in daemon mode.
+
+**Note 2**: The implementation uses Kakoune´s `quit!` command instead of `quit`.
```
Apostrophe is `’`.

useredsa replied:
I mistaked it for `´` (acute accent). The fact that is is called [`Right Single Quotation Mark` ](https://codepoints.net/U+2019) also confused me. It should be fixed now.

alexherbo2 replied:
It looks present everywhere.

useredsa replied:
> It looks present everywhere.

Srry, forgot to add before ammending. Reupdated.

alexherbo2 commented:
@useredsa Would you be ok to present the new features in a post?

I think something is missing in the project, is how the magical works, what is the core of connect.kak and how we build things upon these blocks.

Also about conventions, why connect commands start with **:**.

useredsa commented:
> @useredsa Would you be ok to present the new features in a post?

Instead of in the project repository or in addition? I think the documentation for the project shall be with the project and a separate post can be written in the forum commenting the additions.

> I think something is missing in the project, is how the magical works, what is the core of connect.kak and how we build things upon these blocks.

Yes. My intention is that this information should be in the integration file. Right now it's pretty short, but it could be expanded in future commits, don't you think?



> Also about conventions, why connect commands start with **:**.

Hmm... I was using faq.md to include this kind of things. It is true that instead of frenquently asked questions it's more a file about conventions/design decisions. I vote for changing the name.

alexherbo2 commented:
Yep in Discuss or reddit.

useredsa commented:
Last minute changes:

```
diff --git a/README.md b/README.md
index 6598834..e7a2a0d 100644
--- a/README.md
+++ b/README.md
@@ -217,13 +217,13 @@ PS1='$(~/.local/share/kak/connect/prompt) $ '
 - [Commands]
 - [Aliases]
 - [Modules]
-- [FAQ]
+- [Extra documentation for users][user-extra-documentation]
 - [Recipes]
 - [Integration with other tools][integration]
  
 [Commands]: rc/connect/commands
 [Aliases]: rc/connect/aliases
 [Modules]: rc/modules
-[faq]: docs/faq.md
+[user-extra-documentation]: docs/user-extra-documentation.md
 [recipes]: docs/recipes.md
 [integration]: docs/integration.md
diff --git a/docs/faq.md b/docs/user-extra-documentation.md
similarity index 74%
rename from docs/faq.md
rename to docs/user-extra-documentation.md
index cd6bc36..9348726 100644
--- a/docs/faq.md
+++ b/docs/user-extra-documentation.md
@@ -1,6 +1,8 @@
-# FAQ
+# Extra documentation for users
 
-## What’s the difference between `>` and `$`?
+## FAQ
+
+### What’s the difference between `>` and `$`?
 
 `>` or `connect-terminal` runs its argument
 (which by default is your shell)
@@ -11,7 +13,7 @@ Some programs must be run inside a terminal,
 while others spawn their own graphical window.
 Use `>` for the first type and `$` for the second.
 
-## What’s the purpose of `&` or `connect-detach`?
+### What’s the purpose of `&` or `connect-detach`?
 
 `connect-detach` detaches (closes) the current client
 (leaving you in a terminal if you launched Kakoune there) and
@@ -24,7 +26,7 @@ you didn’t start it in daemon mode.
 
 **Note 2**: The implementation uses Kakoune’s `quit!` command instead of `quit`.
 
-## What is `kak-shell`?
+### What is `kak-shell`?
 
 `kak-shell` is inspired by [nix-shell].
 
@@ -52,7 +54,11 @@ For example, you could run `kak-shell tig` to run [tig] connected to an arbitrar
 [tig]: https://github.com/jonas/tig
 [nix-shell]: https://nixos.org/nix/manual#sec-nix-shell
 
-## What’s the difference between [commands] and [modules]?
+---
+
+## Conventions and design decisions
+
+### What’s the difference between [commands] and [modules]?
 
 The term module is used for small program integrations for Kakoune
 (which are implemented as Kakoune’s modules).
@@ -61,3 +67,11 @@ The term command is used for small binaries that can be used by a connected appl
 [commands]: ../rc/connect/commands
 [modules]:  ../rc/modules
 
+### Why do all [commands] start with a colon (`:`)?
+
+The colon is the key used to access the prompt menu inside Kakoune.
+Because the commands usually resemble Kakoune commands,
+prepending them with a colon is a good way to reuse our muscle memory,
+and it makes it very simple to differentiate them from traditional commands
+(think of `pwd` versus `:pwd`).
+

```

useredsa commented:
> Yep in Discuss or reddit.

If you want me to post it I prefer to do it in discuss, since I don't use reddit. I'm fine if you prefer to post it yourself.

alexherbo2 commented:
I’m not good for explaining things. xd

useredsa commented:
> I’m not good for explaining things. xd

:open_mouth: :facepalm: ...
